### PR TITLE
LOC report: ignore type-only TypeScript churn

### PR DESCRIPTION
**Pull request by mmkal**
## What changes

The PR LOC report keeps raw TypeScript churn in **Lines**, but omits source lines with no runtime
output from **Significant**. TypeScript emits a source map; that map acts as a stencil over the
original source, so the compiler decides which lines have runtime meaning without changing their
count.

```ts
type User = {
  id: string
  displayName: string // Lines +1; Significant +0
}

enum Direction {
  Up, // still Significant: enums emit JavaScript
}
```

This avoids a home-grown TypeScript syntax parser. Interfaces, aliases, declaration files, type-only
imports, and type-only members disappear; enums and other runtime-emitting syntax remain. Source
maps also avoid inflated counts from compiler expansion.

## Proof

- Task/spec committed first in `tasks/ci-loc-report-ignore-type-only-lines.md`.
- Six real-git integration tests cover type-only source and declaration files, mixed TSX, enums,
  JS/docs, and rendered-report behavior.
- `pnpm --dir scripts typecheck` and all 275 scripts workspace tests pass.
- At `4e5e7fc29`, this branch reported **+250/−19 Lines** vs **+216/−11 Significant**.

## Real-world dry run: PR #2301

Running this reporter against [PR #2301](https://github.com/iterate/iterate/pull/2301), from
`1366879` to `bb9cc2f`:

| Group | Lines | Significant |
| --- | ---: | ---: |
| Product | +263 −174 | +187 −139 |
| Tests | +258 −150 | +250 −149 |
| Docs | +48 −0 | +36 −0 |
| **Total** | **+569 −324** | **+473 −288** |

Its current report shows **+529/−309 Significant**. The new filter excludes another 56 added and 21
removed type-only lines while leaving raw Lines unchanged; 54/20 of those are Product code.

## Environment Config Lease

<details>
<summary>No preview slot recorded.</summary>

| app | status | commit | preview | size (gzip) | deploy duration | test duration | retries | cleanup duration | workflow run | updated | summary |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Auth | released | `524a12d` | [https://auth.iterate-preview-3.com](https://auth.iterate-preview-3.com) | 811.7 KiB | 23.0s | 12.6s |  | 3.4s | [Workflow run](https://github.com/iterate/iterate/actions/runs/197612783304303) | 2026-07-24T13:37:55.407Z | Preview app released. |
| Dummy Petshop | released | `524a12d` | [https://dummy-petshop.iterate-preview-3.com](https://dummy-petshop.iterate-preview-3.com) | 198.3 KiB | 6.7s | 12.6s |  | 1.0s | [Workflow run](https://github.com/iterate/iterate/actions/runs/197612783304303) | 2026-07-24T13:37:53.044Z | Preview app released. |
| OS | released | `524a12d` | [https://os.iterate-preview-3.com](https://os.iterate-preview-3.com) | 4.89 MiB (-10.1 KiB vs main) | 69.5s | 88.1s |  | 15.6s | [Workflow run](https://github.com/iterate/iterate/actions/runs/197612783304303) | 2026-07-24T13:38:07.642Z | Preview app released. |
| Semaphore | released | `524a12d` | [https://semaphore.iterate-preview-3.com](https://semaphore.iterate-preview-3.com) | 441.1 KiB | 17.0s | 19.1s |  | 897ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/197612783304303) | 2026-07-24T13:37:52.898Z | Preview app released. |
| Streams Example App | released | `524a12d` | [https://streams.iterate-preview-3.com](https://streams.iterate-preview-3.com) | 1.44 MiB | 67.5s | 60.8s |  | 8.4s | [Workflow run](https://github.com/iterate/iterate/actions/runs/197612783304303) | 2026-07-24T13:38:00.369Z | Preview app released. |

</details>

| Group | Lines | Significant |
| --- | ---: | ---: |
| Tests | +195 -0 | +180 -0 🟩🟩🟩🟩🟩 |
| CI & scripts | +64 -19 | +54 -11 🟩🟩⬜⬜⬜ |
| Docs | +66 -0 | +55 -0 🟩🟩⬜⬜⬜ |
| Generated | +3 -0 | +3 -0 🟩⬜⬜⬜⬜ |
| **Total** | **+328 -19** | **+292 -11** 🟩🟩🟩🟩🟩 |

<sub>Lines counts every changed line; Significant ignores blank lines, JS comments, and TypeScript lines with no runtime output. Between 1366879 and e7d0f2b, bucketed first-match-wins into the groups defined in `scripts/ci/loc-report.ts`.</sub>

---

> [!NOTE]
> **Low Risk**
> CI-only metric logic in `scripts/ci`; no product runtime or auth paths, with regression tests covering edge cases.
> 
> **Overview**
> The PR LOC report still shows every changed line in **Lines**, but **Significant** now drops TypeScript that does not emit runtime JavaScript—interfaces, type-only edits, and declaration files—so reviewers see less type churn without losing raw diff size.
> 
> `scripts/ci/loc-report.ts` transpiles `.ts`/`.tsx`/`.mts`/`.cts` with source maps, keeps only source lines mapped to emitted output (via `@jridgewell/sourcemap-codec`), and treats `.d.ts`/`.d.mts`/`.d.cts` as wholly type-only. JS comment stripping and non-TS behavior are unchanged. `getChangedFiles` takes a `cwd` for git and tests; the PR-body footnote documents the new rule.
> 
> Adds `scripts/ci/loc-report.test.ts` (six real-git fixtures) and `@jridgewell/sourcemap-codec` in the scripts workspace; a completed task note records the spec.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e7d0f2bde0c247abcdf5191909eed944aaa5e3c1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: scripts/ci/loc-report.ts [hunk 1]
```diff
@@ -132,19 +141,44 @@ function linguistGeneratedPaths(paths: string[]): Set<string> {
...
+    const transpiled = ts.transpileModule(content, {
+      fileName: path,
+      compilerOptions: {
+        jsx: ts.JsxEmit.Preserve,
+        module: ts.ModuleKind.ESNext,
+        sourceMap: true,
+        target: ts.ScriptTarget.ESNext,
+      },
+    });
+    if (!transpiled.sourceMapText) throw new Error(`TypeScript emitted no source map for ${path}`);
```
[the remark this answers is not in the document]
Comment (mmkal) [reply scripts/ci/loc-report.ts:160]: 🤖 Fixed in 66786d388. Declaration files now short-circuit to an empty Significant representation before `transpileModule`; a real-git integration case covers `.d.ts`, `.d.mts`, and `.d.cts` and proves each remains raw +1 / Significant +0. Typecheck and all 275 scripts tests pass.

File: scripts/ci/loc-report.ts [hunk 2]
```diff
@@ -132,19 +141,45 @@ function linguistGeneratedPaths(paths: string[]): Set<string> {
...
+    const sourceMap = JSON.parse(transpiled.sourceMapText) as { mappings: string };
```
[the remark this answers is not in the document]
Comment (mmkal) [reply scripts/ci/loc-report.ts]: 🤖 Fixed in 5472bb1dd. The cast is gone: source-map JSON is parsed as `unknown`, then its object shape and string `mappings` field are validated before decode. Malformed compiler output now fails with a clear path-specific error. Focused tests and scripts typecheck pass.
